File: 704-binary-search/binary-search-upper-bound.java
class Solution {
    public int search(int[] nums, int target) {
        if (nums.length == 0) {
            return -1;
        }

        int l = 0;
        int m = nums.length / 2;
        int r = nums.length;

        while (l < r) {
            if (nums[m] <= target) {
                l = m + 1;
            } else {
                r = m;
            }

            m = (l + r) / 2;
        }

        if (l > 0 && nums[l - 1] == target) {
            return l - 1;
        } else {
            return -1;
        }
    }
}

File: BinarySearch/704-binary-search/binary-search-upper-bound.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if (nums.empty())
            return -1;
    
        size_t l = 0;
        size_t m = nums.size() / 2;
        size_t r = nums.size();

        while (l < r) {
            if (nums[m] <= target)
                l = m + 1;
            else
                r = m;
            
            m = (l + r) / 2;
        }

        if (l > 0 && nums[(l - 1)] == target)
            return l - 1;
        else
            return -1;
    }
};